Gem/Code/Source/Weapons/BaseWeapon.cpp

```diff
@@ -88,24 +123,108 @@ namespace MultiplayerSample
         m_fireParams = fireParams;
     }
 
-    const ClientEffect& BaseWeapon::GetActivateEffect() const
+    void BaseWeapon::ExecuteActivateEffect([[maybe_unused]] const AZ::Transform& activateTransform, [[maybe_unused]] const AZ::Vector3& target) const
     {
-        return m_activateEffect;
-    }
+#if AZ_TRAIT_CLIENT
+        if (m_activateEffect != nullptr)
+        {
+            if (PopcornFX::PopcornFXRequests* popcornFx = PopcornFX::PopcornFXRequestBus::FindFirstHandler())
+            {
+                popcornFx->EffectSetTransform(m_activateEffect, activateTransform);
 
-    const ClientEffect& BaseWeapon::GetImpactEffect() const
-    {
-        return m_impactEffect;
+                int32_t lengthAttrId = popcornFx->EffectGetAttributeId(m_activateEffect, "Max Length");
+                if (lengthAttrId >= 0)
+                {
+                    const float length = target.GetDistance(activateTransform.GetTranslation());
+                    popcornFx->EffectSetAttributeAsFloat(m_activateEffect, lengthAttrId, length);
+                }
+
+                int32_t hitAttrId = popcornFx->EffectGetAttributeId(m_activateEffect, "Hit Position");
+                if (hitAttrId >= 0)
+                {
+                    popcornFx->EffectSetAttributeAsFloat3(m_activateEffect, hitAttrId, target);
+                }
+
+                popcornFx->EffectRestart(m_activateEffect, true);
+            }
```

Comment:
(Repeated) If there's any chance that this could run on a thread, you might want to convert this to execute a lambda on the bus instead. To convert, you'd basically just replace the first if statement with something like this:
```
PopcornFX::PopcornFXRequestBus::Broadcast(
    [=](PopcornFX::PopcornFXRequestBus::Events* popcornFx)
    {
...
```
The main difference is that it will keep the bus mutex locked while executing the code inside the lambda, so there's no concern of the popcornFx handler getting removed during execution.

Reply:
(Repeated) Also, I'm not 100% loving these hard-coded attribute names, since these can be changed in the particle effect setups right? I almost wonder if they should be exposed on the component as well.

Reply:
No chance of it being threaded at least given the current entity/component design, but I 100% agree with the hard-coded attribute thing. There are script-canvas bindings for emfx that could also potentially make this a lot more elegant?

Reply:
Done